### ISSUE TITLE
Consider using GCC

### ISSUE THREAD

**Issue by andrew-m-leonard** [adoptium/temurin-build#4091]
GCC with ELF and Binutils supports generation of "build-id" sections for additional debugging information for use when using separate debuginfo: https://sourceware.org/gdb/current/onlinedocs/gdb.html/Separate-Debug-Files.html

Consider whether the Adoptium GCC DevKit build should use --enable-linker-build-id

**Comment by jerboaa** [comment]
From the [gnu linker man page](https://linux.die.net/man/1/ld):

```
--build-id
--build-id=style
    Request creation of ".note.gnu.build-id" ELF note section. The contents of the note are unique bits identifying this linked file. style can be "uuid" to use 128 random bits, "sha1" to use a 160-bit SHA1 hash on the normative parts of the output contents, "md5" to use a 128-bit MD5 hash on the normative parts of the output contents, or "0x hexstring " to use a chosen bit string specified as an even number of hexadecimal digits ("-" and ":" characters between digit pairs are ignored). If style is omitted, "sha1" is used.

    The "md5" and "sha1" styles produces an identifier that is always the same in an identical output file, but will be unique among all nonidentical output files. It is not intended to be compared as a checksum for the file's contents. A linked file may be changed later by other tools, but the build ID bit string identifying the original linked file does not change.

    Passing "none" for style disables the setting from any "--build-id" options earlier on the command line. 
```

**Comment by jerboaa** [comment]
Specifically this:

```
The "md5" and "sha1" styles produces an identifier that is always the same in an identical output file.
```

**Comment by jerboaa** [comment]
Two random builds don't produce identical build-id's for `libjvm.so` (for builds without any of the reproducible build options) on Linux.

**Comment by jerboaa** [comment]
@andrew-m-leonard So I've gone through the motions of:

1. Get to a state where I can produce binary identical builds using the adoptium devkit (of JDK 21). I've produced 3 binary identical builds, because I wasn't convinced two will do ;-)
2. Do the same with a different devkit that has the `--enable-linker-build-id` option. Do again 3 builds.

Here are the results.

## Adoptium devkit
```
$ ls
jdk21u-build-1  jdk21u-build-2  jdk21u-build-3
$ ./jdk21u-build-1/bin/java -Xinternalversion
OpenJDK 64-Bit Server VM (21.0.9+9-LTS) for linux-amd64 JRE (21.0.9+9-LTS), built on 2025-10-21T00:00:00Z by "admin" with gcc 11.3.0
$ diff -r jdk21u-build-1 jdk21u-build-2
$ diff -r jdk21u-build-1 jdk21u-build-3
$ readelf -n jdk21u-build-3/lib/server/libjvm.so | grep -A3 \.note\.gnu\.build-id
```

**Result:** OK, so we have 3 identical builds but no `.note.gnu.debug-id` in the `libjvm.so` binary.

## Devkit with `--enable-linker-build-id` enabled for the GCC build
```
$ ls
jdk21u-build-1  jdk21u-build-2  jdk21u-build-3
$ ./jdk21u-build-1/bin/java -Xinternalversion
OpenJDK 64-Bit Server VM (21.0.9+9-LTS) for linux-amd64 JRE (21.0.9+9-LTS), built on 2025-10-21T00:00:00Z by "admin" with gcc 11.3.0
$ diff -r jdk21u-build-1 jdk21u-build-2
$ diff -r jdk21u-build-1 jdk21u-build-3
$ readelf -n jdk21u-build-3/lib/server/libjvm.so | grep -A3 \.note\.gnu\.build-id
Displaying notes found in: .note.gnu.build-id
  Owner                Data size 	Description
  GNU                  0x00000014	NT_GNU_BUILD_ID (unique build ID bitstring)
    Build ID: b5ec14b2d11f7f674f2bfa3b034b6f271914aa6d
$ readelf -n jdk21u-build-1/lib/server/libjvm.so | grep -A3 \.note\.gnu\.build-id
Displaying notes found in: .note.gnu.build-id
  Owner                Data size 	Description
  GNU                  0x00000014	NT_GNU_BUILD_ID (unique build ID bitstring)
    Build ID: b5ec14b2d11f7f674f2bfa3b034b6f271914aa6d
$ readelf -n jdk21u-build-2/lib/server/libjvm.so | grep -A3 \.note\.gnu\.build-id
Displaying notes found in: .note.gnu.build-id
  Owner                Data size 	Description
  GNU                  0x00000014	NT_GNU_BUILD_ID (unique build ID bitstring)
    Build ID: b5ec14b2d11f7f674f2bfa3b034b6f271914aa6d
```

**Result:** OK, so we have 3 identical builds, *with* `.note.gnu.debug-id` in binaries. Specifically the `Build ID` of `b5ec14b2d11f7f674f2bfa3b034b6f271914aa6d` is the same across all 3 JDK builds with that devkit for `libjvm.so`.

## Conclusion

`--enable-linker-build-id` has no effect on reproducibility. While builds using devkits *with* `--enable-linker-build-id` and builds using devkits that *don't use it* aren't comparable/reproducible, individual builds using the same devkit also **with**  `--enable-linker-build-id` remain reproducible.

**Comment by andrew-m-leonard** [comment]
Thank you @jerboaa , so this looks like it is feasible for Temurin to adopt build-id's, which would be good.
It would however mean "Comparable" builds, between different providers, eg.RH vs Temurin, would need a "fix".

**Comment by jerboaa** [comment]
> It would however mean "Comparable" builds, between different providers, eg.RH vs Temurin, would need a "fix".

@andrew-m-leonard Why? Right now RH builds use build-ids, Temurin don't. So it would be a fix removing that difference.

**Comment by andrew-m-leonard** [comment]
@jerboaa yeah sorry, that's what I meant, as in adding would be the fix, RH vs Temurin Compare is currently broken without the fix!

**Comment by andrew-m-leonard** [comment]
i've added it to the Q4 plan

**Comment by jerboaa** [comment]
@andrew-m-leonard Are we planning to include this for the January updates?

**Comment by andrew-m-leonard** [comment]
> [@andrew-m-leonard](https://github.com/andrew-m-leonard) Are we planning to include this for the January updates?

@jerboaa Possibly, if I can complete it before I leave end of next week.

Presumably this is just jdk-21+ ?

We have DevKits for jdk21(gcc 11.3.0) and jdk25+(gcc 14.2.0)
Architectures: x64, aarch64, s390x and ppc64le. Does it apply to all?

I presume we just need to add "--enable-linker-build-id" to about here ?: https://github.com/openjdk/jdk21u/blob/3c0530fd5f9615941144d8ac89786ce699027b51/make/devkit/Tools.gmk#L464

**Comment by jerboaa** [comment]
> > [@andrew-m-leonard](https://github.com/andrew-m-leonard) Are we planning to include this for the January updates?
> 
> [@jerboaa](https://github.com/jerboaa) Possibly, if I can complete it before I leave end of next week.
> 
> Presumably this is just jdk-21+ ?

Ideally anything that builds with a devkit as the resulting build would support `perf` usage to be run on the JVM.

> We have DevKits for jdk21(gcc 11.3.0) and jdk25+(gcc 14.2.0) Architectures: x64, aarch64, s390x and ppc64le. Does it apply to all?

Yes. I would think that it would apply to all arches for both devkits.

> I presume we just need to add "--enable-linker-build-id" to about here ?: https://github.com/openjdk/jdk21u/blob/3c0530fd5f9615941144d8ac89786ce699027b51/make/devkit/Tools.gmk#L464

I'm not sure how the Temurin Devkits are being built. If it's the plain makefiles from OpenJDK then yes, I think so.

**Comment by judovana** [comment]
> I'm not sure how the Temurin Devkits are being built. If it's the plain makefiles from OpenJDK then yes, I think so.

Sorry for slightly off topic question, but I'm also struggling how the Temurin Devkits are built. IS th ere any doc or repo for them?

**Comment by jerboaa** [comment]
> > I'm not sure how the Temurin Devkits are being built. If it's the plain makefiles from OpenJDK then yes, I think so.
> 
> Sorry for slightly off topic question, but I'm also struggling how the Temurin Devkits are built. IS th ere any doc or repo for them?

It's off-topic, but now that you mention it I was pointed to this at some point, which should help:
https://adoptium.net/en-GB/docs/reproducible-verification-builds/reproduce-linux-x64

**Comment by andrew-m-leonard** [comment]
@judovana Doc not sure!, But basically:
- DevKit build scripts : https://github.com/adoptium/ci-jenkins-pipelines/tree/master/pipelines/build/devkit
- ci.adoptium.net DevKit build jobs : https://ci.adoptium.net/job/build-scripts/job/utils/job/devkit/

**Comment by jerboaa** [comment]
Noting here that the JDK `21.0.10+5` EA build doesn't yet have the chance (built/released around Dec 5, while the fix landed Dec 8). So for JDK 21 EA builds to see changes it needs to be `21.0.10+6` or better.

Similarly for JDK 25u. The last Linux EA build is from November 25. So the same applies there.

**Comment by jerboaa** [comment]
> Similarly for JDK 25u. The last Linux EA build is from November 25. So the same applies there.

Update: Found a JDK 25u build from CI.
```
openjdk 25.0.3-beta 2026-04-21
OpenJDK Runtime Environment Temurin-25.0.3+0-202512151229 (build 25.0.3-beta-202512151229)
OpenJDK 64-Bit Server VM Temurin-25.0.3+0-202512151229 (build 25.0.3-beta-202512151229, mixed mode, sharing)
```

That one supports Linux `perf` probes as described in #4223 (i.e. has debug-id support).
